crs_application/src/crs_application.cpp

```diff
@@ -0,0 +1,139 @@
+static const std::string CURRENT_ST_TOPIC = "current_state";
+static const std::string EXECUTE_ACTION_SERVICE = "execute_action";
+static const std::string GET_AVAILABLE_ACTIONS_SERVICE = "get_available_actions";
+
+int main(int argc, char** argv)
+{
+  using namespace crs_application;
+  using namespace crs_msgs;
+  using namespace scxml_core;
+
+  rclcpp::init(argc, argv);
+  std::shared_ptr<rclcpp::Node> node = std::make_shared<rclcpp::Node>(NODE_NAME, rclcpp::NodeOptions());
+
+  // crs executive
+  CRSExecutive exec(node);
+
+  // qt application
+  QApplication app(argc, argv);
+
+  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr state_pub_ = node->create_publisher<std_msgs::msg::String>(CURRENT_ST_TOPIC,
+                                                                                                      rclcpp::QoS(rclcpp::KeepLast(7)));
+
+  rclcpp::Service<srv::ExecuteAction>::SharedPtr exec_action_server = node->create_service<srv::ExecuteAction>(
+      EXECUTE_ACTION_SERVICE,[&exec, &node](std::shared_ptr<srv::ExecuteAction::Request> req,
+          std::shared_ptr<srv::ExecuteAction::Response> res) -> void{
+
+    res->succeeded = true;
+    if(exec.getSM()->isBusy())
+    {
+      res->succeeded = false;
+      res->err_msg = "State Machine is busy";
+      RCLCPP_ERROR(node->get_logger(), res->err_msg.c_str());
+      return;
+    }
+
+    Response sm_res = exec.getSM()->execute(Action{id : req->action_id, data : boost::any()});
+    if(!sm_res)
+    {
+      res->succeeded = false;
+      res->err_msg = sm_res.msg;
+      RCLCPP_ERROR(node->get_logger(), res->err_msg.c_str());
+      return;
+    }
+  });
+
+  rclcpp::Service<srv::GetAvailableActions>::SharedPtr get_actions_srv = node->create_service<srv::GetAvailableActions>(
+      GET_AVAILABLE_ACTIONS_SERVICE,[&exec, &node](std::shared_ptr<srv::GetAvailableActions::Request> req,
+          std::shared_ptr<srv::GetAvailableActions::Response> res) -> void{
+
+    res->succeeded =false;
+    if(exec.getSM()->isBusy())
+    {
+      res->succeeded = false;
+      res->err_msg = "State Machine is busy";
+      RCLCPP_ERROR(node->get_logger(), res->err_msg.c_str());
+      return;
+    }
+    std::vector<std::string> action_ids = exec.getSM()->getAvailableActions();
+    if(req->search_pattern.empty())
+    {
+      res->action_ids = action_ids;
+      return;
+    }
+
+    if(action_ids.empty())
+    {
+      return;
+    }
+
+    for(const auto& id : action_ids)
+    {
+      if(std::regex_search(id, std::regex(req->search_pattern)))
+      {
+        res->action_ids.push_back(id);
+      }
+    }
+    RCLCPP_WARN_EXPRESSION(node->get_logger(),res->action_ids.empty(),"No action ids matched the '%s' search pattern",
+                           req->search_pattern.c_str());
+  });
+
+  // main loop
+  rclcpp::Rate throttle(100);
+  while(rclcpp::ok())
+  {
+    app.processEvents(QEventLoop::AllEvents);
+    throttle.sleep();
+    rclcpp::spin_some(node);
+  }
+
```

Comment:
Does app require any shutdown?

Reply:
as a shutdown state? or qt?

Reply:
Yeah. Do you need to call anything to shutdown your qt app? Since it's not a gui there may not be.

Reply:
It does not